core/src/main/java/org/mule/runtime/core/internal/config/CustomService.java

```diff
@@ -18,28 +23,31 @@
  *
  * @since 4.0
  */
-public class CustomService {
+public class CustomService<T> {
 
-  private Optional<Class> serviceClass;
-  private Optional<Object> serviceImpl;
+  private final String serviceId;
+  private final Optional<Class> serviceClass;
```

Comment:
unneficient use of optional. Except very few cases, it's generally not a good practice to have Optional fields. Just keep the class and consumer and handle null internally.